tidy3d/plugins/autograd/invdes/filters.py

```diff
@@ -1,71 +1,226 @@
-from functools import partial
-from typing import Tuple, Union
+import abc
+from functools import lru_cache, partial
+from typing import Callable, Iterable, Tuple, Union
 
 import numpy as np
+import pydantic.v1 as pd
+from numpy.typing import NDArray
+
+from tidy3d.components.base import Tidy3dBaseModel
 
 from ..functions import convolve
 from ..types import KernelType, PaddingType
-from ..utilities import make_kernel
+from ..utilities import get_kernel_size_px, make_kernel
+
+
+class AbstractFilter(Tidy3dBaseModel, abc.ABC):
+    """A filter class for creating and applying convolution filters.
+
+    Parameters
+    ----------
+    kernel_size : Tuple[pd.PositiveInt, ...]
+        Size of the kernel in pixels for each dimension.
+    normalize : bool = True
+        Whether to normalize the kernel so that it sums to 1.
+    padding : PaddingType = "reflect"
+        The padding mode to use.
+    """
+
+    kernel_size: Union[pd.PositiveInt, Tuple[pd.PositiveInt, ...]] = pd.Field(
+        ..., description="Size of the kernel in pixels for each dimension."
+    )
+    normalize: bool = pd.Field(
+        True, description="Whether to normalize the kernel so that it sums to 1."
+    )
+    padding: PaddingType = pd.Field("reflect", description="The padding mode to use.")
+
+    @classmethod
+    def from_radius_dl(
+        cls, radius: Union[float, Tuple[float, ...]], dl: Union[float, Tuple[float, ...]], **kwargs
+    ) -> "AbstractFilter":
+        """Create a filter from radius and grid spacing.
+
+        Parameters
+        ----------
+        radius : Union[float, Tuple[float, ...]]
+            The radius of the kernel. Can be a scalar or a tuple.
+        dl : Union[float, Tuple[float, ...]]
+            The grid spacing. Can be a scalar or a tuple.
+        **kwargs
+            Additional keyword arguments to pass to the filter constructor.
+
+        Returns
+        -------
+        AbstractFilter
+            An instance of the filter.
+        """
+        kernel_size = get_kernel_size_px(radius=radius, dl=dl)
+        return cls(kernel_size=kernel_size, **kwargs)
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_kernel(size_px: Iterable[int], normalize: bool) -> NDArray:
+        """Get the kernel for the filter.
+
+        Parameters
+        ----------
+        size_px : Iterable[int]
+            Size of the kernel in pixels for each dimension.
+        normalize : bool
+            Whether to normalize the kernel so that it sums to 1.
+
+        Returns
+        -------
+        NDArray
+            The kernel.
+        """
+        ...
```

Comment:
what does this do? like `pass`?

Reply:
Yeah it's functionally identical. I like using `pass` to convey that something is not supposed to do anything and an ellipsis if something is to be implemented (like a todo in a sense). Hence `...` in method stubs for abstract classes. But I can change it if you want.

Reply:
no strong preference, mostly curious